### PR TITLE
Harden authorization code nonce validation

### PR DESCRIPTION
**Pull request by jo-arroyo**
## Summary

- Enforce bidirectional nonce validation for authorization code token responses
- Reject and warn when an ID token contains a nonce but no expected nonce was supplied
- Preserve optional nonce behavior when neither the request nor ID token contains a nonce
- Document the fail-closed behavior and cover matching, missing, mismatched, empty, and unsolicited nonce cases

## Validation

- `@azure/msal-common`: build, lint, tests, format check, API Extractor
- `@azure/msal-node`: build, lint, tests, format check, API Extractor

assistance: agentic-cli
type: security
agent-tool: copilot-cli
agent-model: gpt-5.6-sol
work-item: AB#n/a

### CODE REVIEW

File: lib/msal-common/src/response/ResponseHandler.ts [hunk 1]
```diff
@@ -219,15 +219,24 @@ export class ResponseHandler {
...
+                );
+            }
+
+            if ((expectedNonce || tokenNonce) && expectedNonce !== tokenNonce) {
+                throw createClientAuthError(
+                    ClientAuthErrorCodes.nonceMismatch,
+                    request.correlationId
+                );
             }
         }
```
[the remark this answers is not in the document]
Comment (jo-arroyo) [reply lib/msal-common/src/response/ResponseHandler.ts]: Addressed in 1491216a3. Nonce presence now uses explicit `!== undefined` checks, so empty-string nonce values are treated as present. MSAL Node also preserves an explicitly empty request nonce, and the tests cover both an empty unsolicited token nonce and a missing token nonce with an empty expected nonce. The strict presence check is scoped to Node authorization-code redemption so browser hybrid-SPA flows retain their existing behavior.

File: lib/msal-common/test/response/ResponseHandler.spec.ts [hunk 2]
```diff
@@ -170,6 +170,169 @@ describe("ResponseHandler.ts", () => {
...
+            await expect(
+                responseHandler.handleServerTokenResponse(
+                    AUTHENTICATION_RESULT.body,
+                    testAuthority,
+                    TimeUtils.nowSeconds(),
+                    testRequest,
+                    0,
+                    {
+                        code: TEST_TOKENS.AUTHORIZATION_CODE,
+                        nonce: undefined,
```
Comment (lalimasharda) [inline lib/msal-common/test/response/ResponseHandler.spec.ts:263]: Wouldn't this mean if nonce supplied as `nonce: null`, it will be evaluated as nonce is present? Is that by design?

File: lib/msal-node/src/client/ClientApplication.ts [hunk 3]
```diff
@@ -177,13 +177,15 @@ export abstract class ClientApplication {
...
-            validatedAuthCodePayload = {
-                ...validatedAuthCodePayload,
-                code: request.code,
-                nonce: request.nonce,
-            };
-        }
+        const expectedNonce =
+            request.nonce !== undefined
+                ? request.nonce
+                : validatedAuthCodePayload?.nonce;
```
Comment (sameerag) [inline lib/msal-node/src/client/ClientApplication.ts:183]: Why is it defaulting to `validatedAuthCodePayload` when it is not set until the next line?

File: lib/msal-common/test/response/ResponseHandler.spec.ts (hunk 2 above)
Comment (jo-arroyo) [reply lib/msal-common/test/response/ResponseHandler.spec.ts:263]: Ah, missed that edge case. Thanks for the catch, updated to catch non-string values.

File: lib/msal-node/src/client/ClientApplication.ts (hunk 3 above)
Comment (jo-arroyo) [reply lib/msal-node/src/client/ClientApplication.ts:183]: `validatedAuthCodePayload` was initialized earlier on line 164 with `authCodePayload`.

File: lib/msal-common/src/response/ResponseHandler.ts [hunk 4]
```diff
@@ -219,15 +219,38 @@ export class ResponseHandler {
...
+        if (
+            authCodePayload &&
+            Object.prototype.hasOwnProperty.call(authCodePayload, "nonce")
+        ) {
+            const expectedNonce = authCodePayload.nonce;
+            const tokenNonce = idTokenClaims?.nonce;
+
+            // Warn when the ID Token nonce cannot be bound to the request.
+            if (tokenNonce !== undefined && expectedNonce === undefined) {
+                this.logger.warning(
```
Comment (konstantin-msft) [inline lib/msal-common/src/response/ResponseHandler.ts:237]: Logger message says "Rejecting the response", but the block itself does not throw. Do i miss anything?

File: lib/msal-node/test/client/PublicClientApplication.spec.ts [hunk 5]
```diff
@@ -224,6 +245,10 @@ describe("PublicClientApplication", () => {
         await authApp.acquireTokenByCode(request);
 
         expect(AuthorizationCodeClient).toHaveBeenCalledTimes(1);
+        expect(acquireTokenSpy.mock.calls[0][2]).toEqual({
```
Comment (konstantin-msft) [inline lib/msal-node/test/client/PublicClientApplication.spec.ts]: `toEqual` ignores undefined keys. `toStrictEqual` is a better alternative

File: lib/msal-common/src/response/ResponseHandler.ts (hunk 4 above)
Comment (jo-arroyo) [reply lib/msal-common/src/response/ResponseHandler.ts:237]: It would throw in the next block, but I added the throw here too to make it easier to read. Thanks!

File: lib/msal-node/test/client/PublicClientApplication.spec.ts (hunk 5 above)
Comment (jo-arroyo) [reply lib/msal-node/test/client/PublicClientApplication.spec.ts]: Good point, updated. Thanks!

**Review by konstantin-msft — APPROVED** [review]
